### ISSUE TITLE
Collection functions to "pick" and "omit" keys from a map

### ISSUE THREAD

**Issue by coen-hyde** [hashicorp/terraform#29415]
### Current Terraform Version
```
1.0.3
```

### Use-cases
I would like a Collection functions to do the functional equivalent of https://lodash.com/docs/4.17.15#pick and https://lodash.com/docs/4.17.15#omit. ie compose a new map from an existing map, keeping the structure identical but selectively manage which keys are in the new map based on a provided key list. 

### Attempted Solutions
I haven't found an elegant way to do this using Terraform's native functions. Please let me know if I have missed something or you have a simple solution.

### Proposal
Implement equivalents of https://lodash.com/docs/4.17.15#pick and https://lodash.com/docs/4.17.15#omit.

**Comment by remilapeyre** [comment]
Hi @coen-hyde, have you tried

```hcl
{
    for k, v in var.foo:  k => v 
    if contains(var.keep, k) 
}
```

?

**Comment by coen-hyde** [comment]
@realflash I have not. I made things way more complicated than that. Thanks for the suggestion. 

I guess this request has less necessity now, but it would still be nice to have a native function.

**Comment by apparentlymart** [comment]
Hi @coen-hyde! Thanks for sharing this use-case.

I see in the discussion here that you've already found a way to get this done with existing functions. Given that this hasn't been a common request so far, my initial instinct is that the pair of existing features that @remilapeyre proposed compose to produce something good enough for this occasional situation, and `for` also generalizes to arbitrary filter expressions that can meet other use-cases beyond just key matching.

Given that, I'm going to label this as an enhancement so it can stay on our radar for future consideration, but I also want to be up front that it's likely to have a relatively low priority since we tend to prioritize adding and maintaining features that address situations that Terraform can't already support, or that can have a broad impact across multiple use-cases.

Thanks again!

**Comment by crw** [comment]
Thank you for your continued interest in this issue. 

[Terraform version 1.8](https://github.com/hashicorp/terraform/blob/v1.8.0-beta1/CHANGELOG.md) launches with support of provider-defined functions. It is now possible to implement your own functions! We would love to see this implemented as a provider-defined function. 

Please see the [provider-defined functions documentation](https://developer.hashicorp.com/terraform/plugin/framework/functions) to learn how to implement functions in your providers. If you are new to provider development, learn how to [create a new provider](https://developer.hashicorp.com/terraform/plugin/framework) with the Terraform Plugin Framework. If you have any questions, please visit the [Terraform Plugin Development category in our official forum](https://discuss.hashicorp.com/c/terraform-providers/tf-plugin-sdk/43). 

We hope this feature unblocks future function development and provides more flexibility for the Terraform community. Thank you for your continued support of Terraform!
